File: options/typess/cash.py
import datetime

from dataclasses import dataclass
from typing import Dict

from options.typess.enums import TickType, Resolution
from options.typess.security import Security, SecurityDataSnap
from options.typess.scenario import Scenario


@dataclass
class Cash(Security):
    symbol: str
    multiplier: int = 1

    def __post_init__(self):
        self.symbol = self.symbol.upper()

    @property
    def underlying_symbol(self) -> str:
        return self.symbol

    @classmethod
    def from_filename(cls, filename: str):
        """hpe_quote.csv"""
        symbol = filename
        return cls(symbol)

    def csv_name(self, tick_type: TickType, resolution: Resolution, date: datetime.date = None):
        raise NotImplementedError

    def zip_name(self, tick_type: TickType, resolution: Resolution, date: datetime.date = None):
        raise NotImplementedError

    @staticmethod
    def delta(*args, **kwargs):
        return 1

    @staticmethod
    def iv(*args, **kwargs):
        return 0

    @staticmethod
    def npv():
        return 1

    def nlv(self, market_data: Dict[str, SecurityDataSnap], q: float = 1, scenario=Scenario.mid):
        return self.multiplier * q

    def __repr__(self):
        return f'{self.symbol.upper()}'

    def __hash__(self):
        return hash(self.__repr__())

    def __str__(self):
        return self.__repr__()

File: fino/src/types/Cash.jl
# types/Cash.jl

using Dates

"""
    Cash <: Security

Represents a cash position (currency).

Fields:
- `symbol::String` — currency code (stored uppercase), e.g. "USD"
- `multiplier::Int` — contract multiplier (default 1)
"""
    mutable struct Cash <: Security
    symbol::String
    multiplier::Int

    function Cash(symbol::String; multiplier::Int = 1)
        new(uppercase(symbol), multiplier)
    end
end

# ── Implement Security interface ───────────────────────────────────────────────

function underlying_symbol(cash::Cash)::String
    return cash.symbol
end

function csv_name(cash::Cash, tick_type, resolution, dt::Union{Date, Nothing} = nothing)::String
    throw(NotImplementedError("csv_name not implemented for Cash"))
end

function zip_name(cash::Cash, tick_type, resolution, dt::Union{Date, Nothing} = nothing)::String
    throw(NotImplementedError("zip_name not implemented for Cash"))
end

function nlv(cash::Cash, market_data::Dict, q::Float64 = 1.0, scenario::String = "mid")::Float64
    return cash.multiplier * q
end

# ── Greek approximations (cash returns constant values) ────────────────────────

"""
    delta(cash::Cash, args...; kwargs...)

Cash has delta = 1 (moves dollar-for-dollar).
"""
function delta(cash::Cash, args...; kwargs...)::Float64
    return 1.0
end

"""
    iv(cash::Cash, args...; kwargs...)

Cash has no implied volatility (return 0).
"""
function iv(cash::Cash, args...; kwargs...)::Float64
    return 0.0
end

"""
    npv(cash::Cash, args...; kwargs...)

Net present value for cash is 1.0 (normalized).
"""
function npv(cash::Cash, args...; kwargs...)::Float64
    return 1.0
end

# ── String representation & hashing ────────────────────────────────────────────

    Base.repr(cash::Cash) = uppercase(cash.symbol)

    Base.string(cash::Cash) = repr(cash)

    Base.hash(cash::Cash) = hash(repr(cash))

# ── Factory method ────────────────────────────────────────────────────────────

"""
    cash_from_filename(filename::String)::Cash

Create a Cash from a filename (e.g., "USD_quote.csv").
Extracts symbol from the filename.
"""
function cash_from_filename(filename::String)::Cash
    symbol = split(filename, "_")[1]
    return Cash(symbol)
end